ls1intum/Artemis
# `General`: Simplify LTI configuration for instructors

### Checklist
#### General
#### Server
#### Client

### Motivation and Context
1. We want to make setting up an LTI course as easy as possible for an instructor. Right now, the information for how to setup an LTI exercise is spread out over all exercises in the course. We want to move it all to one place, and make it more clear for instructors what each fields is used for through the use of tooltips.
2. When creating/editing an online course it's not very clear what all the fields are used for, and how they should be filled in by the instructor.
3. When setting up an LTI exercise on another system (e.g. Moodle), it can be annoying to have to copy a lot of different fields manually.
4. When an LTI user is initialized, they are presented with a checkbox explaining their account was created and a password was created as well. They should save the password or they will have to reset it. It's easy for users to click ok without actually reading the dialog.

### Description
1. Moved the LTI configuration page from the exercise page to the course page, split the fields into different tabs to make it clearer what fields belong together and added tooltips. Add edit button to edit configuration. Added tooltips to help the user understand the meaning of each field. Moved the validation logic in the server from CourseResource to OnlineCourseConfigurationService.
2. Removed all the LTI fields from the course edit/create page, course can only be set as online course or not when editing course, and configuration is created with default values.
3. Added a copy button to make copying to clipboard easier.
4. When an LTI user is initialized, they are presented with a checkbox explaining their account was created and a password was created as well. They should save the password or they will have to reset it. It's easy for users to click ok without actually reading the dialog.

### Steps for Testing

Prerequisites:
- 1 Instructor

Course creation:

1. Log in to Artemis
2. Create/edit a course
3. Verify that the LTI fields are not present anymore
4. Verify that creating an online course creates a default configuration
5. Verify that it works to make a non-online course an online course
5. Verify that it works to an online course a non-online course

LTI configuration page:

1. Log in to Artemis
2. Create an online course
3. Navigate to the LTI Configuration page by clicking the "LtiConfiguration" link in the course page
4. Verify that the tabs work and that all information is shown
6. Verify that the copy buttons work
7. Verify that the LTI configuration page is no longer available for exercises

Edit LTI configuration page:

1. Log in to Artemis
2. Create an online course
3. Navigate to the LTI Configuration page by clicking the "LtiConfiguration" link in the course page
4. Verify that clicking the edit button takes you to a new page
5. Verify that the fields have tooltips and are translated
6. Verify that the fields that are required must be filled in
8. Verify that the user prefix field cannot have special characters
9. Verify that the save and cancel button work

### Review Progress

#### Code Review
#### Manual Tests

### Test Coverage
| Class/File | Branch | Line |
|------------|-------:|-----:|
| OnlineCourseConfiguration.java | 100% | 95% |
| CourseResource.java | 73% | 93% |
| LtiResource.java | 75% | 95% |
| OnlineCourseConfigurationService.java | 85% | 100% |
| LtiDynamicRegistrationService.java | 88% | 100% |
| course-management.service.ts | 90% | 100% |
| course-lti-configuration.component.ts | 76% | 100% |
| edit-course-lti-configuration.component.ts | 76% | 100% |
| copy-icon.button.component.ts | 100% | 100% |

### Screenshots

####  LTI initialization checkbox

##### Before
![ltiUserInitialization](https://user-images.githubusercontent.com/11006274/196517846-186e9ba7-a44b-4cad-99e1-2117e8f54b9a.png)

##### After

![ltiUserInitAfter](https://user-images.githubusercontent.com/11006274/196519531-f6d023f7-d84d-40e1-bb9b-aec8979da575.png)

####  Online Course creation

##### Before

https://user-images.githubusercontent.com/11006274/202268368-9d2589bd-db10-4402-ad43-12bcddbfba3a.mp4

##### After

https://user-images.githubusercontent.com/11006274/202268938-a96506c7-843c-4016-b16f-7e7d7b0baf50.mp4


####  LTI configuration page

##### Before

https://user-images.githubusercontent.com/11006274/200122142-6397afe4-a54c-4d00-8f1a-5b7289cc0d1a.mp4

##### After


https://user-images.githubusercontent.com/11006274/202269714-ad4e3889-331e-4c7d-955f-7c223a73d003.mp4
